File: packages/esm-format/src/validate.test.ts
/**
 * Tests for structural validation
 */

import { describe, it, expect } from 'vitest';
import { validate } from './validate.js';

describe('Structural validation', () => {
  it('should detect equation count mismatch', () => {
    const data = {
      esm: "0.1.0",
      metadata: { name: "test" },
      models: {
        TestModel: {
          variables: {
            x: { type: "state", default: 1.0 },
            y: { type: "state", default: 2.0 }
          },
          equations: [
            { lhs: { op: "D", args: ["x"], wrt: "t" }, rhs: "y" }
          ]
        }
      }
    };

    const result = validate(data);

    expect(result.is_valid).toBe(false);
    expect(result.structural_errors).toHaveLength(1);
    expect(result.structural_errors[0].code).toBe('equation_count_mismatch');
    expect(result.structural_errors[0].path).toBe('/models/TestModel');
    expect(result.structural_errors[0].details.missing_equations_for).toEqual(['y']);
    expect(result.unit_warnings).toBeDefined();
    expect(Array.isArray(result.unit_warnings)).toBe(true);
  });

  it('should detect undefined variable in equation', () => {
    const data = {
      esm: "0.1.0",
      metadata: { name: "test" },
      models: {
        TestModel: {
          variables: {
            x: { type: "state", default: 1.0 }
          },
          equations: [
            { lhs: { op: "D", args: ["x"], wrt: "t" }, rhs: "undefined_var" }
          ]
        }
      }
    };

    const result = validate(data);

    expect(result.is_valid).toBe(false);
    expect(result.structural_errors.some(err => err.code === 'undefined_variable')).toBe(true);
    expect(result.unit_warnings).toBeDefined();
    expect(Array.isArray(result.unit_warnings)).toBe(true);
  });

  it('should detect undefined system in coupling', () => {
    const data = {
      esm: "0.1.0",
      metadata: { name: "test" },
      models: {
        TestModel: {
          variables: {
            x: { type: "state", default: 1.0 }
          },
          equations: [
            { lhs: { op: "D", args: ["x"], wrt: "t" }, rhs: 1.0 }
          ]
        }
      },
      coupling: [
        {
          type: "operator_compose",
          systems: ["TestModel", "NonExistentModel"]
        }
      ]
    };

    const result = validate(data);

    expect(result.is_valid).toBe(false);
    expect(result.structural_errors.some(err => err.code === 'undefined_system')).toBe(true);
    expect(result.unit_warnings).toBeDefined();
    expect(Array.isArray(result.unit_warnings)).toBe(true);
  });

  it('should detect null reaction', () => {
    const data = {
      esm: "0.1.0",
      metadata: { name: "test" },
      reaction_systems: {
        TestSystem: {
          species: {
            A: { default: 1.0 }
          },
          parameters: {
            k: { default: 0.1 }
          },
          reactions: [
            {
              id: "R1",
              substrates: null,
              products: null,
              rate: "k"
            }
          ]
        }
      }
    };

    const result = validate(data);

    expect(result.is_valid).toBe(false);
    expect(result.structural_errors.some(err => err.code === 'null_reaction')).toBe(true);
    expect(result.unit_warnings).toBeDefined();
    expect(Array.isArray(result.unit_warnings)).toBe(true);
  });

  it('should detect undefined species in reaction', () => {
    const data = {
      esm: "0.1.0",
      metadata: { name: "test" },
      reaction_systems: {
        TestSystem: {
          species: {
            A: { default: 1.0 }
          },
          parameters: {
            k: { default: 0.1 }
          },
          reactions: [
            {
              id: "R1",
              substrates: [{ species: "B", stoichiometry: 1 }], // B not declared
              products: [{ species: "A", stoichiometry: 1 }],
              rate: "k"
            }
          ]
        }
      }
    };

    const result = validate(data);

    expect(result.is_valid).toBe(false);
    expect(result.structural_errors.some(err => err.code === 'undefined_species')).toBe(true);
    expect(result.unit_warnings).toBeDefined();
    expect(Array.isArray(result.unit_warnings)).toBe(true);
  });

  it('should pass validation for valid data', () => {
    const data = {
      esm: "0.1.0",
      metadata: { name: "test" },
      models: {
        TestModel: {
          variables: {
            x: { type: "state", default: 1.0 },
            y: { type: "parameter", default: 2.0 }
          },
          equations: [
            { lhs: { op: "D", args: ["x"], wrt: "t" }, rhs: "y" }
          ]
        }
      }
    };

    const result = validate(data);

    expect(result.is_valid).toBe(true);
    expect(result.structural_errors).toHaveLength(0);
    expect(result.unit_warnings).toBeDefined();
    expect(Array.isArray(result.unit_warnings)).toBe(true);
  });

  it('should include unit_warnings field in ValidationResult', () => {
    const data = {
      esm: "0.1.0",
      metadata: { name: "test" },
      models: {
        TestModel: {
          variables: {
            x: { type: "state", default: 1.0 }
          },
          equations: [
            { lhs: { op: "D", args: ["x"], wrt: "t" }, rhs: 1.0 }
          ]
        }
      }
    };

    const result = validate(data);

    // Ensure the ValidationResult has all required fields per spec Section 3.4
    expect(result).toHaveProperty('is_valid');
    expect(result).toHaveProperty('schema_errors');
    expect(result).toHaveProperty('structural_errors');
    expect(result).toHaveProperty('unit_warnings');
    expect(Array.isArray(result.unit_warnings)).toBe(true);
  });
});

File: packages/esm-format-go/pkg/esm/validate_test.go
package esm

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestValidateValidModel(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestModel",
			Authors: []string{"Test Author"},
		},
		Models: map[string]Model{
			"TestModel": {
				Variables: map[string]ModelVariable{
					"x": {
						Type:    "state",
						Units:   strPtr("m"),
						Default: 0.0,
					},
					"y": {
						Type:       "observed",
						Expression: "x",
					},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.True(t, result.Valid)
	assert.Empty(t, result.Messages)
}

func TestValidateModelWithUnknownVariable(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestModel",
			Authors: []string{"Test Author"},
		},
		Models: map[string]Model{
			"TestModel": {
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: "unknown_var", // This variable doesn't exist
					},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.False(t, result.Valid)
	assert.Len(t, result.Messages, 1)
	assert.Contains(t, result.Messages[0].Message, "Unknown variable 'unknown_var'")
	assert.Equal(t, "error", result.Messages[0].Level)
}

func TestValidateObservedVariableWithoutExpression(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestModel",
			Authors: []string{"Test Author"},
		},
		Models: map[string]Model{
			"TestModel": {
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
					"y": {Type: "observed"}, // Missing expression
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.False(t, result.Valid)
	assert.Len(t, result.Messages, 1)
	assert.Contains(t, result.Messages[0].Message, "Observed variable must have an expression")
}

func TestValidateReactionSystem(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestReactions",
			Authors: []string{"Test Author"},
		},
		ReactionSystems: map[string]ReactionSystem{
			"TestReactions": {
				Species: map[string]Species{
					"A": {Units: strPtr("mol/mol")},
					"B": {Units: strPtr("mol/mol")},
				},
				Parameters: map[string]Parameter{
					"k": {Units: strPtr("1/s")},
				},
				Reactions: []Reaction{
					{
						ID:         "R1",
						Substrates: []SubstrateProduct{{Species: "A", Stoichiometry: 1}},
						Products:   []SubstrateProduct{{Species: "B", Stoichiometry: 1}},
						Rate:       "k",
					},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.True(t, result.Valid)
	assert.Empty(t, result.Messages)
}

func TestValidateReactionWithUnknownSpecies(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestReactions",
			Authors: []string{"Test Author"},
		},
		ReactionSystems: map[string]ReactionSystem{
			"TestReactions": {
				Species: map[string]Species{
					"A": {Units: strPtr("mol/mol")},
				},
				Parameters: map[string]Parameter{
					"k": {Units: strPtr("1/s")},
				},
				Reactions: []Reaction{
					{
						ID:         "R1",
						Substrates: []SubstrateProduct{{Species: "A", Stoichiometry: 1}},
						Products:   []SubstrateProduct{{Species: "UnknownSpecies", Stoichiometry: 1}},
						Rate:       "k",
					},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.False(t, result.Valid)
	assert.Len(t, result.Messages, 1)
	assert.Contains(t, result.Messages[0].Message, "Unknown species 'UnknownSpecies'")
}

func TestValidateComplexExpression(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestModel",
			Authors: []string{"Test Author"},
		},
		Models: map[string]Model{
			"TestModel": {
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
					"y": {Type: "state"},
					"k": {Type: "parameter"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: ExprNode{
							Op: "*",
							Args: []interface{}{
								"k",
								ExprNode{Op: "+", Args: []interface{}{"x", "y"}},
							},
						},
					},
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"y"}, Wrt: strPtr("t")},
						RHS: ExprNode{
							Op: "*",
							Args: []interface{}{
								"k",
								ExprNode{Op: "-", Args: []interface{}{"x", "y"}},
							},
						},
					},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.True(t, result.Valid)
	assert.Empty(t, result.Messages)
}

func TestValidateDiscreteEvent(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestModel",
			Authors: []string{"Test Author"},
		},
		Models: map[string]Model{
			"TestModel": {
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
				DiscreteEvents: []DiscreteEvent{
					{
						Trigger: DiscreteEventTrigger{
							Type:       "condition",
							Expression: ExprNode{Op: ">", Args: []interface{}{"x", 10.0}},
						},
						Affects: []AffectEquation{
							{
								LHS: "x",
								RHS: float64(0.0),
							},
						},
					},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.True(t, result.Valid)
	assert.Empty(t, result.Messages)
}

func TestValidateDataLoaders(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestModel",
			Authors: []string{"Test Author"},
		},
		Models: map[string]Model{
			"TestModel": {
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
			},
		},
		DataLoaders: map[string]DataLoader{
			"TestLoader": {
				Type:     "gridded_data",
				LoaderID: "test_loader",
				Provides: map[string]ProvidedVar{
					"temperature": {Units: "K"},
				},
			},
		},
	}

	result := Validate(esmFile)
	assert.True(t, result.Valid)
	assert.Empty(t, result.Messages)
}

func TestValidateDataLoaderMissingRequiredFields(t *testing.T) {
	esmFile := &EsmFile{
		Esm: "0.1.0",
		Metadata: Metadata{
			Name:    "TestModel",
			Authors: []string{"Test Author"},
		},
		Models: map[string]Model{
			"TestModel": {
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
			},
		},
		DataLoaders: map[string]DataLoader{
			"BadLoader": {
				// Missing Type and LoaderID
				Provides: map[string]ProvidedVar{},
			},
		},
	}

	result := Validate(esmFile)
	assert.False(t, result.Valid)

	// Should have errors for missing type and loader_id, and warning for no provides
	errorCount := 0
	warningCount := 0
	for _, msg := range result.Messages {
		if msg.Level == "error" {
			errorCount++
		} else if msg.Level == "warning" {
			warningCount++
		}
	}

	assert.GreaterOrEqual(t, errorCount, 2) // Type and LoaderID missing
	assert.GreaterOrEqual(t, warningCount, 1) // No variables provided
}

// Test equation-unknown balance validation
func TestValidateEquationUnknownBalance(t *testing.T) {
	tests := []struct {
		name          string
		model         Model
		expectedValid bool
		expectedError string
	}{
		{
			name: "balanced model with one state variable and one ODE",
			model: Model{
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
			},
			expectedValid: true,
		},
		{
			name: "balanced model with two state variables and two ODEs",
			model: Model{
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
					"y": {Type: "state"},
					"k": {Type: "parameter"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: ExprNode{Op: "*", Args: []interface{}{"k", "y"}},
					},
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"y"}, Wrt: strPtr("t")},
						RHS: ExprNode{Op: "*", Args: []interface{}{"k", "x"}},
					},
				},
			},
			expectedValid: true,
		},
		{
			name: "unbalanced model with state variable but no ODE",
			model: Model{
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
					"y": {Type: "state"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
			},
			expectedValid: false,
			expectedError: "state variables without ODE equations: [y]",
		},
		{
			name: "unbalanced model with ODE for non-state variable",
			model: Model{
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
					"k": {Type: "parameter"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"k"}, Wrt: strPtr("t")},
						RHS: float64(2.0),
					},
				},
			},
			expectedValid: false,
			expectedError: "ODE equations for non-state variables: [k]",
		},
		{
			name: "unbalanced model with no state variables but ODEs",
			model: Model{
				Variables: map[string]ModelVariable{
					"k": {Type: "parameter"},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"k"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
				},
			},
			expectedValid: false,
			expectedError: "found 0 state variables but 1 ODE equations",
		},
		{
			name: "model with non-derivative equations (should be balanced)",
			model: Model{
				Variables: map[string]ModelVariable{
					"x": {Type: "state"},
					"y": {Type: "observed", Expression: ExprNode{Op: "*", Args: []interface{}{"x", 2.0}}},
				},
				Equations: []Equation{
					{
						LHS: ExprNode{Op: "D", Args: []interface{}{"x"}, Wrt: strPtr("t")},
						RHS: float64(1.0),
					},
					{
						// This is not an ODE, it's an algebraic constraint
						LHS: "y",
						RHS: ExprNode{Op: "*", Args: []interface{}{"x", 2.0}},
					},
				},
			},
			expectedValid: true,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			esmFile := &EsmFile{
				Esm: "0.1.0",
				Metadata: Metadata{
					Name:    "TestModel",
					Authors: []string{"Test Author"},
				},
				Models: map[string]Model{
					"TestModel": tc.model,
				},
			}

			result := Validate(esmFile)

			assert.Equal(t, tc.expectedValid, result.Valid, "Validation result should match expected")

			if !tc.expectedValid {
				assert.NotEmpty(t, result.Messages, "Should have validation messages")

				found := false
				for _, msg := range result.Messages {
					if tc.expectedError != "" && assert.Contains(t, msg.Message, tc.expectedError) {
						found = true
						break
					}
				}

				if tc.expectedError != "" {
					assert.True(t, found, "Should find expected error message containing: %s", tc.expectedError)
				}
			} else {
				// Check that there are no equation-unknown balance errors
				for _, msg := range result.Messages {
					assert.NotContains(t, msg.Message, "Equation-unknown balance", "Should not have equation-unknown balance errors")
				}
			}
		})
	}
}

func TestValidateFileSpecCompliant(t *testing.T) {
	// Test the new spec-compliant ValidateFile function
	jsonStr := `{
		"esm": "0.1.0",
		"metadata": {
			"name": "Test",
			"authors": ["Test Author"]
		},
		"models": {
			"TestModel": {
				"variables": {
					"x": {"type": "state", "default": 0.0},
					"y": {"type": "state", "default": 0.0}
				},
				"equations": [
					{
						"lhs": {"op": "D", "args": ["x"], "wrt": "t"},
						"rhs": "y"
					}
				]
			}
		}
	}`

	esmFile, err := LoadString(jsonStr)
	require.NoError(t, err)

	result := ValidateFile(esmFile, jsonStr)

	// Check that result has the correct structure per spec
	assert.NotNil(t, result)
	assert.NotNil(t, result.SchemaErrors)
	assert.NotNil(t, result.StructuralErrors)
	assert.NotNil(t, result.UnitWarnings)

	// Schema should be valid
	assert.Empty(t, result.SchemaErrors, "No schema errors expected for valid JSON")

	// Should have structural error due to equation-unknown balance (2 state vars, 1 ODE equation)
	assert.NotEmpty(t, result.StructuralErrors, "Should have structural error for equation count mismatch")
	assert.False(t, result.IsValid, "Should be invalid due to structural errors")

	// Check that structural error has proper code
	if len(result.StructuralErrors) > 0 {
		foundEquationError := false
		for _, err := range result.StructuralErrors {
			if err.Code == ErrorEquationCountMismatch {
				foundEquationError = true
				assert.Contains(t, err.Message, "Equation-unknown balance")
			}
		}
		assert.True(t, foundEquationError, "Should have equation count mismatch error")
	}
}

func TestValidateFileValidModel(t *testing.T) {
	// Test with a properly balanced model
	jsonStr := `{
		"esm": "0.1.0",
		"metadata": {
			"name": "Test",
			"authors": ["Test Author"]
		},
		"models": {
			"TestModel": {
				"variables": {
					"x": {"type": "state", "default": 0.0},
					"k": {"type": "parameter", "default": 1.0}
				},
				"equations": [
					{
						"lhs": {"op": "D", "args": ["x"], "wrt": "t"},
						"rhs": {"op": "*", "args": ["k", "x"]}
					}
				]
			}
		}
	}`

	esmFile, err := LoadString(jsonStr)
	require.NoError(t, err)

	result := ValidateFile(esmFile, jsonStr)

	// Should be valid - 1 state variable, 1 ODE equation
	assert.Empty(t, result.SchemaErrors)
	assert.Empty(t, result.StructuralErrors)
	assert.True(t, result.IsValid)
}